### ISSUE TITLE
Add API endpoint for content-addressed downloads

### ISSUE THREAD

**Issue by mathstuf** [girder/girder#961]
Midas allows one to download using `&checksum=$hash&algorithm=$algo`. It would be nice to have such content-addressed download URLs for girder as well (something like `item/content/{algo}/{hash}/download`?).

These URLs are used for VTK's data store downloading (where the `{id}` of the file is not necessarily known).

Of course, if there is a better way, I'm willing to be enlightened :) .

**Comment by mathstuf** [comment]
Cc: @brad.king

**Comment by jcfr** [comment]
Per discussion with @zachmullen , a possible approach will be to develop a `ContentAddressedDownload` plugin allowing to:
* enable the content-addressed download on per-collection basis
* configure which hashing algorithm should be associated with a given collection
* batch-update existing item in a given collection based on the selected hash algorithm

Notes:
* At first, since downloading data from cloud asset-store is usually not free, the plugin will be available only if a filesystem asset-store is enabled. If there is a need, cloud based asset-store support could be added later.

**Comment by brianhelba** [comment]
I have a need for this now, and after discussing it with @jbeezley , I'm planning to implement this as a plugin where:
* Upon upload of a file to any non-S3 assetstore type, the plugin will hash the file with all of the algorithms that [CMake ExternalData supports](https://cmake.org/cmake/help/v3.3/module/ExternalData.html#hash-algorithms).
 * Since non-proxied uploads to S3 never give Girder direct custody of the file's binary data, we can't guarantee full support for S3 assetstores. However, if a proxied S3 upload takes case, the plugin should hash the content. Finally, note that it's possible to [get the MD5 (but not other hashes) of a file in S3 without downloading it](http://stackoverflow.com/questions/1775816/how-to-get-the-md5sum-of-a-file-on-amazons-s3)
 * Generating multiple hash types shouldn't be too much more expensive, since we can have multiple stateful hash variables, each of which is updated as the file is iterated through once.
* The file may be downloaded via a ``/api/v1/file/content/{algo}/{hash}/download`` URL.
 * I don't have a strong opinion about the syntax of this URL, so chime in if you have a better idea.
* At first, this plugin will not support access to files that were uploaded before the plugin was enabled.
* The plugin will be enabled for all files globally.
 * I don't really understand the need to only enable this on a per-collection basis. Feel free to present a use case that requires this, however.

**Comment by brianhelba** [comment]
I'm hoping to get this enabled on ``data.kitware.com`` ASAP, so I'll try to get it ready in time for the next release.

**Comment by vovythevov** [comment]
I'll be working on this.

**Comment by brianhelba** [comment]
Here is the code where the SHA512 hash is currently generated for uploads to a FilesystemAssetstore: https://github.com/girder/girder/blob/master/girder/utility/filesystem_assetstore_adapter.py#L118

**Comment by vovythevov** [comment]
@brianhelba: I am hitting a (rather big) bump. The problem is to be able to serialize the hashes (hash?). The python hashlib library doesn't offer these sorts of capabilities.
@zachmullen implemented it for sha512 based on http://code.activestate.com/recipes/578479-save-and-restore-sha-512-internal-state/.  
I haven't been able to find anything on other types (beside some small clues for md5: https://groups.google.com/forum/#!topic/comp.lang.python/5TDP_0j2ET8). 
What did you have in mind to be able to serialize the hashes' states ?

**Comment by brianhelba** [comment]
@vovythevov Check out #1226 , that should take care of it.

**Comment by opadron** [comment]
Sorry if this had already been addressed, but how would this download interact with the access controls?  If the hash refers to a file I don't have access to, can I still download it using its hash?  What if it *also* refers to a shallow copy that I *do* have access to?

**Comment by zachmullen** [comment]
Once we move this into girder, the logic will be just like it is in Midas. 

```
files = fetch all files from DB with matching checksum
foreach file in files
    if user has access to file
        return file
raise exception('no matching file')
```

**Comment by jcfr** [comment]
Could we consider the following use case:
* Private commercial project 
* Hosting private data on https://data.kitware.com
* Using CMake module `ExternalData`

How will the authentication work ? 

Can we create an application token that will be used in the fetch URL associated with the CMake ExternalData module ?

**Comment by zachmullen** [comment]
Yes.

**Comment by zachmullen** [comment]
@jcfr #1197 is designed as a first step toward explicit token management. Once that's fleshed out, it would support cases like this. If this is a feature you want to see sooner rather than later, we should talk.

**Comment by opadron** [comment]
@zachmullen Would be nice to preferentially return the file from a local assetstore over, say, hdfs or S3.

**Comment by zachmullen** [comment]
@opadron why would that be?

**Comment by jcfr** [comment]
If it would be possible to have data stored in multiple asset-store, preferring the one where data are hosted locally can make sense. Download would be faster. In our case, this would apply to https://data.kitware.com 

That said, I don't think there is a way to have multiple asseststore enabled ...

**Comment by zachmullen** [comment]
That only makes sense if your client is known to be closer to the girder server than, say, S3, by some distance metric. For S3 downloads, the client downloads the data directly from S3, which could actually be a lot faster depending on the environment your tests are running in.

**Comment by brianhelba** [comment]
This is a reasonable point: regardless of which assetstore type we prefer, we probably do want to have a deterministic way of picking. So, the code will look something like:

```
def getFileByHash(hashType, hashValue, user=None)
    filesByAssetstore = {
        # If multiple files are in the same assetstore, it doesn't matter which is used
        file['assetstoreId']: file
        for file in FileModel.find({'hash.%s' % hashType: hashValue}):
        if FileModel.hasAccess(file, user)
    }
    if not filesByAssetstore:
        # Or raise an exception
        return None
    elif len(filesByAssetstore) == 1:
        return filesByAssetstore.popitem()
    else:
        assetstores = [
            AssetstoreModel.load(assetstoreId)
            for assetstoreId in filesByAssetstore.viewkeys()
        ]
        # TODO: how do we pick an assetstore?
        preferredAssetstore = assetstores[0]
        return filesByAssetstore[preferredAssetstore['_id']
```

**Comment by zachmullen** [comment]
That's fine, but this is an optimization that I don't think needs to be in the first iteration of this plugin. By the way, does this exist somewhere yet?

**Comment by brianhelba** [comment]
@vovythevov is working on it.

I wrote the code just now for the sake of clarity in the discussion.

**Comment by jcfr** [comment]
> # TODO: how do we pick an assetstore?

For now, I suggest assestore selected is the one marked as "current" in the admin console

**Comment by opadron** [comment]
I was just thinking that for a download endpoint, where I figured Girder would fetch the data from whatever assetstore has the file and then pass it onto the client.  I agree that for other scenarios where the client might access the data directly, it's hard to know what assetstore to prefer a priori.

**Comment by brianhelba** [comment]
This now exists as the "Hashsum download" plugin. There are still opportunities to expand the variety of supported hash functions and assetstores, and for fetching file metadata, but we can wait on specific demands for additional features and enhancements to be requested in additional bug reports.

Closing.

**Comment by jcfr** [comment]
:+1:
